lib/HLSL/DxilValidation.cpp

```diff
@@ -1261,7 +1261,7 @@ static void ValidateResourceOffset(CallInst *CI, DXIL::ResourceKind resKind,
           validateOffset(offsets[i]);
       }
     } else {
-      if (!isa<UndefValue>(offsets[i])) {
+      if (i < offsets.size() && !isa<UndefValue>(offsets[i])) {
```

Comment:
It's not very clear to me why we need to do additional bound check here (`i < offsets.size()`). Is it because we can no longer expect `kMaxNumOffsets` to be 3? It might change based on resource kind?

Reply:
It's been so long I don't remember exactly what I was thinking here. You are correct that `kMaxNumOffsets` should no longer be 3 in all cases. It could have needed to be 4 for the existing cases before too. It shouldn't really be a constant either.

I think I'll remove this conditional and change the `kMaxNumOffsets` loop to use the size of `offsets` and add some more tests for cases that have different expected numbers of offsets and different maximum numbers of offsets

Reply:
Futher up in the stack, only the expected offsets are passed into this function as an array ref.  This array size is not always kMaxNumOffsets, so this bound check prevents OOB index on offsets[i].

Reply:
@pow2clk  I think that should be ok: using `offsets.size()` instead of `kMaxNumOffsets` for loop, and removing the bound check here.  I think this was leftover from a version of the code that had an assumed fixed sized array.